sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -669,11 +672,42 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
       val newConditionOpt = conditionOpt match {
         case Some(condition) =>
           val newFilters = additionalConstraints -- splitConjunctivePredicates(condition)
-          if (newFilters.nonEmpty) Option(And(newFilters.reduce(And), condition)) else None
+          if (newFilters.nonEmpty) Option(And(newFilters.reduce(And), condition)) else conditionOpt
         case None =>
           additionalConstraints.reduceOption(And)
       }
-      if (newConditionOpt.isDefined) Join(left, right, joinType, newConditionOpt) else join
+      // Infer filter for left/right outer joins
+      val newLeftOpt = joinType match {
+        case RightOuter if newConditionOpt.isDefined =>
+          val rightConstraints = right.constraints.union(
+            splitConjunctivePredicates(newConditionOpt.get).toSet)
+          val inferredConstraints = ExpressionSet(
+            QueryPlanConstraints.inferAdditionalConstraints(rightConstraints))
+          val leftConditions = inferredConstraints
```

Comment:
can we call `getRelevantConstraints`?

Reply:
I was too thinking to give it another name since it's public now. How about `getInferredConstraints`? I'm good either way, though.

Reply:
We need to also union `constructIsNotNullConstraints`. Thus, let us use `getRelevantConstraints `

Reply:
I think the `constructIsNotNullConstraints` logic does not deal with the "transitive" constraints so we not need to include it here. Instead the "isNotNull" deduction for inferred filters on the null-supplying side is guaranteed by 2 things here:
1) when getting constraints from the preserved side, `constructIsNotNullConstraints` has already been called and will be carried over by `inferAdditionalConstraints` to the null-supplying side;
2) the Filter matching part of `InferFiltersFromConstraints`.
That said, I'm good with the name `getRelevantConstraints` too.

Reply:
I see. I am proposing to add a new helper function `getRelevantConstraints ` in the trait `QueryPlanConstraints`. I would suggest to ignore what we are doing for this specific case. That function could be called in the other cases in the future.

Reply:
I had the same inclination of wrapping the evaluation of `allConstraints` in a helper function too, till I realized that `constructIsNotNullConstraints` would depend on the LogicalPlan node in addition to the input constraints. `constructIsNotNullConstraints` has two parts, one being to deduce is-not-null from null-intolerant expressions, the other to deduce is-null-not from attribute nullability. We could reorganize these pieces into new methods, but I feel like we should wait till we find an actual usage so to figure out what needs to be included in the helper function.

Reply:
This is my proposal
```Scala
val leftConstraints = left.getRelevantConstraints(
  left.constraints
    .union(right.constraints)
    .union(splitConjunctivePredicates(conditionOpt.get).toSet))
val newFilters = reduceConjunctivePredicates(leftConstraints.toSeq)
  .filterNot(left.constraints.contains)
```

Reply:
yeah, let us use `ExpressionSet `

Reply:
Let us leave `allConstraints ` untouched. We should avoid the extra code changes in this PR. We need to use `conf.constraintPropagationEnabled` for the extra constraints introduced  by this PR.

Reply:
I made another commit yesterday. How's it looking now?